File: mts/src/mts/runtimes/direct_api.py
"""Direct API runtime — uses an LLMProvider for generation/revision."""

from __future__ import annotations

from mts.providers.base import LLMProvider
from mts.runtimes.base import AgentOutput, AgentRuntime


class DirectAPIRuntime(AgentRuntime):
    """Agent runtime that calls an LLM provider directly.

    This is the simplest runtime — equivalent to what the experiment
    scripts and SimpleAgentTask do today.
    """

    def __init__(
        self,
        provider: LLMProvider,
        model: str | None = None,
    ) -> None:
        self._provider = provider
        self._model = model

    def generate(
        self,
        prompt: str,
        system: str | None = None,
        schema: dict | None = None,
    ) -> AgentOutput:
        sys_prompt = system or "You are a skilled writer and analyst. Complete the task precisely."
        result = self._provider.complete(
            system_prompt=sys_prompt,
            user_prompt=prompt,
            model=self._model,
        )
        return AgentOutput(
            text=result.text,
            cost_usd=result.cost_usd,
            model=result.model,
        )

    def revise(
        self,
        prompt: str,
        previous_output: str,
        feedback: str,
        system: str | None = None,
    ) -> AgentOutput:
        revision_prompt = (
            f"Revise the following output based on the judge's feedback.\n\n"
            f"## Original Output\n{previous_output}\n\n"
            f"## Judge Feedback\n{feedback}\n\n"
            f"## Original Task\n{prompt}\n\n"
            "Produce an improved version:"
        )
        sys_prompt = system or "You are revising content based on expert feedback. Improve the output."
        result = self._provider.complete(
            system_prompt=sys_prompt,
            user_prompt=revision_prompt,
            model=self._model,
        )
        return AgentOutput(
            text=result.text,
            cost_usd=result.cost_usd,
            model=result.model,
        )

File: ts/src/runtimes/direct-api.ts
/**
 * Direct API runtime — uses an LLMProvider for generation/revision.
 * Port of autocontext/src/autocontext/runtimes/direct_api.py
 */

import type { LLMProvider } from "../types/index.js";
import type { AgentOutput, AgentRuntime } from "./base.js";

export class DirectAPIRuntime implements AgentRuntime {
  readonly name = "DirectAPI";

  constructor(
    private provider: LLMProvider,
    private model?: string,
  ) {}

  async generate(opts: {
    prompt: string;
    system?: string;
    schema?: Record<string, unknown>;
  }): Promise<AgentOutput> {
    const sys =
      opts.system ??
      "You are a skilled writer and analyst. Complete the task precisely.";
    const result = await this.provider.complete({
      systemPrompt: sys,
      userPrompt: opts.prompt,
      model: this.model,
    });
    return {
      text: result.text,
      costUsd: result.costUsd ?? undefined,
      model: result.model ?? undefined,
    };
  }

  async revise(opts: {
    prompt: string;
    previousOutput: string;
    feedback: string;
    system?: string;
  }): Promise<AgentOutput> {
    const revisionPrompt =
      `Revise the following output based on the judge's feedback.\n\n` +
      `## Original Output\n${opts.previousOutput}\n\n` +
      `## Judge Feedback\n${opts.feedback}\n\n` +
      `## Original Task\n${opts.prompt}\n\n` +
      "Produce an improved version:";

    const sys =
      opts.system ??
      "You are revising content based on expert feedback. Improve the output.";
    const result = await this.provider.complete({
      systemPrompt: sys,
      userPrompt: revisionPrompt,
      model: this.model,
    });
    return {
      text: result.text,
      costUsd: result.costUsd ?? undefined,
      model: result.model ?? undefined,
    };
  }
}
